### PR TITLE
Backports 0.17 pr33

### PR DESCRIPTION
**Pull request by UdjinM6**

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -4322,7 +4322,7 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
     return true;
 }
 
-void CWallet::ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& keypool, bool fInternal)
+bool CWallet::ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& keypool, bool fRequestedInternal)
```
Comment (PastaPastaPasta) [inline src/wallet/wallet.cpp:4324]: did fInternal get renamed in an earlier PR?

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -4349,14 +4350,17 @@ void CWallet::ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& keypool, bool fIn
         if (!HaveKey(keypool.vchPubKey.GetID())) {
             throw std::runtime_error(std::string(__func__) + ": unknown key in key pool");
         }
-        if (keypool.fInternal != fInternal) {
+        if (keypool.fInternal != fReturningInternal) {
```
Comment (PastaPastaPasta) [inline src/wallet/wallet.cpp:4352]: what is `use_split_keypool` from?

**Review by PastaPastaPasta — APPROVED** [review]
LGTM, utACK. Two thoughts

**Comment by UdjinM6** [comment]
@PastaPastaPasta Our HD wallet implementation is a mix of (dirty 🙈 ) backports and Dash specific changes (bip39/44 implementation) https://github.com/dashpay/dash/pull/1405 so quite a few parts of code around it are different. There was no split between external and internal keypools in bitcoin core originally, it was just one HD keypool at first so they have to have `use_split_keypool` to ensure backwards compatibility. We skipped that step and went with 2 keypools from the start and as a result we don't need `use_split_keypool` in our code.

**Comment by PastaPastaPasta** [comment]
Would be nice if we could refactor our hd system to be closer to upstream potentially

**Comment by UdjinM6** [comment]
Rebased after #4197

**Review by PastaPastaPasta — APPROVED** [review]
utACK, force push looks good. Builds and passes tests locally

**Pull request merged**